wallet/src/main/scala/org/bitcoins/wallet/internal/RescanHandling.scala

```diff
@@ -0,0 +1,157 @@
+
+  /////////////////////
+  // Private methods
+
+  private def doNeutrinoRescan(
+      startOpt: Option[BlockStamp],
+      endOpt: Option[BlockStamp],
+      addressBatchSize: Int): Future[Unit] = {
+    for {
+      scriptPubKeys <- generateScriptPubKeys(addressBatchSize)
+      blocks <- matchBlocks(scriptPubKeys, endOpt, startOpt)
+      _ <- downloadAndProcessBlocks(blocks)
+      externalGap <- calcAddressGap(HDChainType.External)
+      changeGap <- calcAddressGap(HDChainType.Change)
+      res <- if (externalGap >= AddressGapLimit && changeGap >= AddressGapLimit)
+        pruneUnusedAddresses()
+      else doNeutrinoRescan(startOpt, endOpt, addressBatchSize)
+    } yield res
+  }
+
+  private def pruneUnusedAddresses(): Future[Unit] = {
+    for {
+      addressDbs <- addressDAO.findAll()
+      _ <- addressDbs.foldLeft(FutureUtil.unit) { (prevF, addressDb) =>
+        for {
+          _ <- prevF
+          spendingInfoDbs <- spendingInfoDAO.findByScriptPubKey(
+            addressDb.scriptPubKey)
+          _ <- if (spendingInfoDbs.isEmpty) addressDAO.delete(addressDb)
+          else FutureUtil.unit
+        } yield ()
+      }
+    } yield ()
+  }
+
+  private def calcAddressGap(chainType: HDChainType): Future[Int] = {
+    for {
+      addressDbs <- addressDAO.findAll()
+      addressGap <- addressDbs
+        .filter(_.path.chain.chainType == chainType)
+        .sortBy(_.path.address.index)
+        .foldLeft(Future.successful(0)) { (prevNF, addressDb) =>
+          for {
+            prevN <- prevNF
+            spendingInfoDbs <- spendingInfoDAO.findByScriptPubKey(
+              addressDb.scriptPubKey)
+          } yield {
+            if (spendingInfoDbs.isEmpty) prevN + 1 else 0
+          }
+        }
+    } yield {
+      logger.debug(s"Address gap: $addressGap")
+      addressGap
+    }
+  }
+
+  private def downloadAndProcessBlocks(
+      blocks: Vector[DoubleSha256Digest]): Future[Unit] = {
+    logger.debug(s"Requesting ${blocks.size} block(s)")
+    blocks.foldLeft(Future.successful(())) { (prevF, blockHash) =>
+      val completedF = subscribeForBlockProcessingCompletionSignal(blockHash)
+      for {
+        _ <- prevF
+        _ <- nodeApi.downloadBlocks(Vector(blockHash))
+        _ <- completedF
+      } yield ()
+    }
+  }
+
+  private def matchBlocks(
+      scriptPubKeys: Vector[ScriptPubKey],
+      endOpt: Option[BlockStamp],
+      startOpt: Option[BlockStamp]): Future[Vector[DoubleSha256Digest]] = {
+    val threadPool =
+      Executors.newFixedThreadPool(Runtime.getRuntime.availableProcessors() * 2)
+
+    val blocksF = for {
+      blocks <- getMatchingBlocks(scriptPubKeys, startOpt, endOpt)(
+        ExecutionContext.fromExecutor(threadPool))
+    } yield {
+      blocks.sortBy(_._2).map(_._1.flip)
+    }
+
+    blocksF.onComplete(_ => threadPool.shutdown())
+    blocksF
+  }
+
+  private def generateScriptPubKeys(
+      count: Int): Future[Vector[ScriptPubKey]] = {
+    for {
```

Comment:
maybe this? This should allow us to run things in parallel. Is this going to cause synchronization issues with `getNewAddress()` you were talking about the other day?

```scala

    val receiveSPKsF: Future[Vector[BitcoinAddress]] = {
      1.to(count)
        .foldLeft(Future.successful(Vector.empty[BitcoinAddress])) {
          (prevFuture, _) =>
            for {
              prev <- prevFuture
              address <- getNewAddress()
            } yield prev :+ address
        }
    }

    val changeAddressesF: Future[Vector[BitcoinAddress]] = {
      1.to(count)
        .foldLeft(Future.successful(Vector.empty[BitcoinAddress])) {
          (prevFuture, _) =>
            for {
              prev <- prevFuture
              address <- getNewChangeAddress()
            } yield prev :+ address
        }
    }

    for {
      receive <- receiveSPKsF
      change <- changeAddressesF
    } yield receive.map(_.scriptPubKey) ++ change.map(_.scriptPubKey)
```

Reply:
There's no acute need in running this in parallel